File: go/017.go
func letterCombinations(digits string) []string {
    if len(digits) == 0 {
        return nil
    }
    m := map[byte][]byte{'2': []byte("abc"), '3': []byte("def"), '4': []byte("ghi"),
        '5': []byte("jkl"), '6': []byte("mno"), '7': []byte("pqrs"), '8': []byte("tuv"), '9': []byte("wxzy")}
    ret, size := []string{""}, len(digits)
    for i := 0; i < size; i++ {
        temp, sz := []string{}, len(ret)
        for j := 0; j < sz; j++ {
            ssss, ok := m[digits[i]]
            if !ok {
                return nil
            }
            bytes := []byte(ret[j])
            for _, c := range ssss {
                bytes = append(bytes, c)
                temp = append(temp, string(bytes))
                bytes = bytes[:len(bytes)-1]
            }
        }
        ret = temp
    }
    return ret
}

File: cpp/017.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty())
            return {};
        int size = digits.size();
        vector<string> ret{""};
        map<char, string> m = {{'2',"abc"},{'3',"def"},{'4',"ghi"},{'5',"jkl"},{'6',"mno"},{'7',"pqrs"},{'8',"tuv"},{'9',"wzxy"}};
        for (int i = 0; i < size; ++i) {
            vector<string> temp;
            int sz = ret.size();
            for (int j = 0; j < sz; ++j) {
                string s = m[digits[i]];
                for (int k = 0, s_sz = s.size(); k < s_sz; ++k) {
                    temp.push_back(ret[j] + s[k]);
                }
            }
            ret = temp;
        }
        return ret;
    }
};
